File: Week-Wise-Assignments/Assignment 3/Systematic Solutions/C Programs/2_delete_operations.c
#include <stdio.h>
#include <stdlib.h>

// Node structure for single linked list
typedef struct Node {
    int data;
    struct Node* next;
} Node;

/*
ALGORITHMS FOR DELETE OPERATIONS:

deleteFromFirst Algorithm:
1. Check if list is empty
2. If empty, return NULL
3. Store head in temp
4. Update head to head->next
5. Free temp node
6. Return new head

deleteFromLast Algorithm:
1. Check if list is empty
2. If only one node, delete it and return NULL
3. Traverse to second last node
4. Store last node in temp
5. Set second last node's next to NULL
6. Free temp node

deleteFromAny Algorithm:
1. Check if position is valid
2. If position is 1, call deleteFromFirst
3. Traverse to (position-1)th node
4. Store position node in temp
5. Set (position-1)th node's next to temp->next
6. Free temp node
*/

// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        printf("Memory allocation failed!\n");
        return NULL;
    }
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to create an empty list
Node* createList() {
    printf("Empty linked list created successfully!\n");
    return NULL;
}

// Function to insert at the end (for testing purposes)
Node* insertAtLast(Node* head, int data) {
    Node* newNode = createNode(data);
    if (newNode == NULL) return head;
    
    if (head == NULL) {
        return newNode;
    }
    
    Node* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    
    temp->next = newNode;
    return head;
}

// Function to delete from the beginning
Node* deleteFromFirst(Node* head) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from first.\n");
        return NULL;
    }
    
    Node* temp = head;
    head = head->next;
    printf("Deleted %d from the beginning\n", temp->data);
    free(temp);
    return head;
}

// Function to delete from the end
Node* deleteFromLast(Node* head) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from last.\n");
        return NULL;
    }
    
    // If only one node
    if (head->next == NULL) {
        printf("Deleted %d from the end\n", head->data);
        free(head);
        return NULL;
    }
    
    // Traverse to second last node
    Node* temp = head;
    while (temp->next->next != NULL) {
        temp = temp->next;
    }
    
    Node* lastNode = temp->next;
    printf("Deleted %d from the end\n", lastNode->data);
    temp->next = NULL;
    free(lastNode);
    return head;
}

// Function to delete from any position
Node* deleteFromAny(Node* head, int position) {
    if (head == NULL) {
        printf("List is empty! Cannot delete from position %d.\n", position);
        return NULL;
    }
    
    if (position < 1) {
        printf("Invalid position! Position should be >= 1\n");
        return head;
    }
    
    // If deleting first node
    if (position == 1) {
        return deleteFromFirst(head);
    }
    
    // Traverse to (position-1)th node
    Node* temp = head;
    for (int i = 1; i < position - 1 && temp != NULL; i++) {
        temp = temp->next;
    }
    
    // If position is beyond the list
    if (temp == NULL || temp->next == NULL) {
        printf("Position %d is beyond the list!\n", position);
        return head;
    }
    
    Node* nodeToDelete = temp->next;
    temp->next = nodeToDelete->next;
    printf("Deleted %d from position %d\n", nodeToDelete->data, position);
    free(nodeToDelete);
    return head;
}

// Function to display the list
void displayList(Node* head) {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }
    
    printf("Linked List: ");
    Node* temp = head;
    while (temp != NULL) {
        printf("%d", temp->data);
        if (temp->next != NULL) {
            printf(" -> ");
        }
        temp = temp->next;
    }
    printf(" -> NULL\n");
}

// Function to count nodes in the list
int countNodes(Node* head) {
    int count = 0;
    Node* temp = head;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// Function to free the entire list
void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
    printf("Memory freed successfully!\n");
}

// Function to populate list with sample data
Node* populateList(Node* head) {
    printf("Populating list with sample data: 10, 20, 30, 40, 50\n");
    head = insertAtLast(head, 10);
    head = insertAtLast(head, 20);
    head = insertAtLast(head, 30);
    head = insertAtLast(head, 40);
    head = insertAtLast(head, 50);
    return head;
}

int main() {
    Node* head = NULL;
    int choice, position;
    
    printf("=== Single Linked List - Delete Operations ===\n");
    
    // Create empty list
    head = createList();
    
    do {
        printf("\n--- Menu ---\n");
        printf("1. Delete from First\n");
        printf("2. Delete from Last\n");
        printf("3. Delete from Any Position\n");
        printf("4. Display List\n");
        printf("5. Count Nodes\n");
        printf("6. Populate with Sample Data\n");
        printf("7. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1:
                head = deleteFromFirst(head);
                break;
                
            case 2:
                head = deleteFromLast(head);
                break;
                
            case 3:
                printf("Enter position to delete (1-based): ");
                scanf("%d", &position);
                head = deleteFromAny(head, position);
                break;
                
            case 4:
                displayList(head);
                break;
                
            case 5:
                printf("Number of nodes: %d\n", countNodes(head));
                break;
                
            case 6:
                head = populateList(head);
                displayList(head);
                break;
                
            case 7:
                printf("Exiting...\n");
                freeList(head);
                break;
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    } while (choice != 7);
    
    return 0;
}

File: Week-Wise-Assignments/Assignment 3/Systematic Solutions/Python Programs/2_delete_operations.py
class Node:
    """
    Node class for single linked list
    """
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    """
    Single Linked List implementation with delete operations
    """
    
    def __init__(self):
        """
        Initialize empty linked list
        """
        self.head = None
    
    def createList(self):
        """
        Create an empty linked list
        """
        self.head = None
        print("Empty linked list created successfully!")
    
    def insertAtLast(self, data):
        """
        Insert node at the end (for testing purposes)
        
        Args:
            data: Data to insert
        """
        new_node = Node(data)
        
        if self.head is None:
            self.head = new_node
            return
        
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        
        temp.next = new_node
    
    def deleteFromFirst(self):
        """
        Delete node from the beginning of the list
        
        Returns:
            Data of deleted node or None if list is empty
        """
        if self.head is None:
            print("List is empty! Cannot delete from first.")
            return None
        
        deleted_data = self.head.data
        self.head = self.head.next
        print(f"Deleted {deleted_data} from the beginning")
        return deleted_data
    
    def deleteFromLast(self):
        """
        Delete node from the end of the list
        
        Returns:
            Data of deleted node or None if list is empty
        """
        if self.head is None:
            print("List is empty! Cannot delete from last.")
            return None
        
        # If only one node
        if self.head.next is None:
            deleted_data = self.head.data
            self.head = None
            print(f"Deleted {deleted_data} from the end")
            return deleted_data
        
        # Traverse to second last node
        temp = self.head
        while temp.next.next is not None:
            temp = temp.next
        
        deleted_data = temp.next.data
        temp.next = None
        print(f"Deleted {deleted_data} from the end")
        return deleted_data
    
    def deleteFromAny(self, position):
        """
        Delete node from any position in the list
        
        Args:
            position: Position to delete from (1-based indexing)
            
        Returns:
            Data of deleted node or None if position is invalid
        """
        if self.head is None:
            print(f"List is empty! Cannot delete from position {position}.")
            return None
        
        if position < 1:
            print("Invalid position! Position should be >= 1")
            return None
        
        # If deleting first node
        if position == 1:
            return self.deleteFromFirst()
        
        # Traverse to (position-1)th node
        temp = self.head
        for i in range(1, position - 1):
            if temp is None or temp.next is None:
                break
            temp = temp.next
        
        # If position is beyond the list
        if temp is None or temp.next is None:
            print(f"Position {position} is beyond the list!")
            return None
        
        deleted_data = temp.next.data
        temp.next = temp.next.next
        print(f"Deleted {deleted_data} from position {position}")
        return deleted_data
    
    def deleteByValue(self, value):
        """
        Delete first occurrence of a value
        
        Args:
            value: Value to delete
            
        Returns:
            bool: True if deleted, False if not found
        """
        if self.head is None:
            print(f"List is empty! Cannot delete value {value}.")
            return False
        
        # If first node contains the value
        if self.head.data == value:
            self.head = self.head.next
            print(f"Deleted value {value} from the beginning")
            return True
        
        # Search for the value
        temp = self.head
        while temp.next is not None:
            if temp.next.data == value:
                temp.next = temp.next.next
                print(f"Deleted value {value} from the list")
                return True
            temp = temp.next
        
        print(f"Value {value} not found in the list")
        return False
    
    def displayList(self):
        """
        Display the linked list
        """
        if self.head is None:
            print("List is empty!")
            return
        
        print("Linked List: ", end="")
        temp = self.head
        while temp is not None:
            print(temp.data, end="")
            if temp.next is not None:
                print(" -> ", end="")
            temp = temp.next
        print(" -> None")
    
    def countNodes(self):
        """
        Count the number of nodes in the list
        
        Returns:
            int: Number of nodes
        """
        count = 0
        temp = self.head
        while temp is not None:
            count += 1
            temp = temp.next
        return count
    
    def isEmpty(self):
        """
        Check if the list is empty
        
        Returns:
            bool: True if empty, False otherwise
        """
        return self.head is None
    
    def populateList(self, values):
        """
        Populate list with given values
        
        Args:
            values: List of values to insert
        """
        print(f"Populating list with values: {values}")
        for value in values:
            self.insertAtLast(value)

def demonstrate_delete_operations():
    """
    Demonstrate delete operations
    """
    print("=== Single Linked List - Delete Operations Demonstration ===")
    
    ll = LinkedList()
    ll.createList()
    
    # Populate with sample data
    sample_data = [10, 20, 30, 40, 50]
    ll.populateList(sample_data)
    ll.displayList()
    
    # Test deleteFromFirst
    print("\n1. Testing deleteFromFirst:")
    ll.deleteFromFirst()
    ll.displayList()
    
    # Test deleteFromLast
    print("\n2. Testing deleteFromLast:")
    ll.deleteFromLast()
    ll.displayList()
    
    # Test deleteFromAny
    print("\n3. Testing deleteFromAny:")
    ll.deleteFromAny(2)  # Delete from position 2
    ll.displayList()
    
    # Test deleteByValue
    print("\n4. Testing deleteByValue:")
    ll.deleteByValue(30)  # Delete value 30
    ll.displayList()
    
    # Display statistics
    print(f"\nTotal nodes: {ll.countNodes()}")
    print(f"Is empty: {ll.isEmpty()}")

def interactive_delete_operations():
    """
    Interactive delete operations
    """
    print("\n=== Interactive Delete Operations ===")
    
    ll = LinkedList()
    ll.createList()
    
    while True:
        print("\n--- Menu ---")
        print("1. Delete from First")
        print("2. Delete from Last")
        print("3. Delete from Any Position")
        print("4. Delete by Value")
        print("5. Display List")
        print("6. Count Nodes")
        print("7. Populate with Sample Data")
        print("8. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                ll.deleteFromFirst()
            
            elif choice == 2:
                ll.deleteFromLast()
            
            elif choice == 3:
                position = int(input("Enter position to delete (1-based): "))
                ll.deleteFromAny(position)
            
            elif choice == 4:
                value = int(input("Enter value to delete: "))
                ll.deleteByValue(value)
            
            elif choice == 5:
                ll.displayList()
            
            elif choice == 6:
                print(f"Number of nodes: {ll.countNodes()}")
            
            elif choice == 7:
                sample_data = [10, 20, 30, 40, 50]
                ll.populateList(sample_data)
                ll.displayList()
            
            elif choice == 8:
                print("Exiting...")
                break
            
            else:
                print("Invalid choice! Please try again.")
        
        except ValueError:
            print("Please enter a valid number!")
        except Exception as e:
            print(f"Error: {e}")

def test_edge_cases():
    """
    Test edge cases for delete operations
    """
    print("\n=== Testing Edge Cases ===")
    
    ll = LinkedList()
    
    # Test operations on empty list
    print("\n1. Operations on empty list:")
    ll.deleteFromFirst()
    ll.deleteFromLast()
    ll.deleteFromAny(1)
    ll.deleteByValue(10)
    
    # Test single node deletion
    print("\n2. Single node deletion:")
    ll.insertAtLast(42)
    ll.displayList()
    ll.deleteFromLast()
    ll.displayList()
    
    # Test invalid positions
    print("\n3. Invalid positions:")
    ll.populateList([1, 2, 3])
    ll.displayList()
    ll.deleteFromAny(0)   # Invalid position
    ll.deleteFromAny(-1)  # Invalid position
    ll.deleteFromAny(10)  # Position beyond list
    ll.displayList()
    
    # Test deleting all nodes
    print("\n4. Deleting all nodes:")
    while not ll.isEmpty():
        ll.deleteFromFirst()
        ll.displayList()

def performance_test():
    """
    Performance test for delete operations
    """
    print("\n=== Performance Test ===")
    
    import time
    
    # Test with large list
    ll = LinkedList()
    n = 1000
    
    print(f"Creating list with {n} nodes...")
    start_time = time.time()
    for i in range(n):
        ll.insertAtLast(i)
    creation_time = time.time() - start_time
    
    print(f"List creation time: {creation_time:.4f} seconds")
    print(f"Initial node count: {ll.countNodes()}")
    
    # Test deleteFromFirst performance
    start_time = time.time()
    for i in range(n // 2):
        ll.deleteFromFirst()
    delete_first_time = time.time() - start_time
    
    print(f"Delete from first (500 operations): {delete_first_time:.4f} seconds")
    print(f"Remaining nodes: {ll.countNodes()}")
    
    # Test deleteFromLast performance
    start_time = time.time()
    while not ll.isEmpty():
        ll.deleteFromLast()
    delete_last_time = time.time() - start_time
    
    print(f"Delete from last (500 operations): {delete_last_time:.4f} seconds")
    print(f"Final node count: {ll.countNodes()}")

if __name__ == "__main__":
    # Run demonstration
    demonstrate_delete_operations()
    
    # Test edge cases
    test_edge_cases()
    
    # Performance test
    performance_test()
    
    # Run interactive mode
    interactive_delete_operations()